torchao/quantization/prototype/mixed_precision/scripts/BO_acc.py

```diff
@@ -0,0 +1,419 @@
+from ax.service.ax_client import AxClient, ObjectiveProperties
+import torch.multiprocessing as mp
+from ax.modelbridge.cross_validation import cross_validate
+
+
+# quantize a model based on a given quantization configuration
+def quantize_by_fqn_to_config(model, device, fqn_to_config):
+    it = iter(fqn_to_config.items())
+    while True:
+        try:
+            k1, v1 = next(it)
+            k2, v2 = next(it)
+            fqn = k1[8:]
+            bit_width, groupsize = v1, v2
+
+
+            def filter_fn_sen(child: torch.nn.Module, cur_fqn: str) -> bool:
+                return isinstance(child, torch.nn.Linear) and (fqn in cur_fqn)
+
+            quantize_(
+                model.to(device=device),
+                intN_weight_only(n=bit_width, group_size=groupsize),
+                filter_fn_sen,
+            )
+        except StopIteration:
+            break
+
+# calculate perplexity on wikitext-document, need to support more tasks
+def cal_wikitext_ppl(model, tokenizer, limit=62):
+
+    with torch.no_grad():
+        result = evaluate(
+            HFLM(pretrained=model, tokenizer=tokenizer, batch_size=1),
+            get_task_dict("wikitext"),
+            limit=limit
+        )
+
+    return result["results"]["wikitext"]["word_perplexity,none"]
+
+def cal_model_size(model, fqn_to_config):
+    _sum = 0
+    fqn_cofg_dict = dict()
+
+    it = iter(fqn_to_config.items())
+    while True:
+        try:
+            k1, v1 = next(it)
+            k2, v2 = next(it)
+            bit_width, groupsize = v1, v2
+            bit_zeropoint = 32
+            bit_scale = 8
+            fqn = k1[8:]
+            fqn_cofg_dict[fqn] = (bit_width, groupsize, bit_zeropoint, bit_scale)
+        except StopIteration:
+            break
+
+    for name, parameter in model.named_parameters():
+        flag = 0
+        for fqn in fqn_cofg_dict:
+            if fqn in name:
+                flag = 1
+                if "self_attn" in name or "mlp" in name:
+                    _sum += parameter.numel() * fqn_cofg_dict[fqn][
+                        0
+                    ] + parameter.numel() // fqn_cofg_dict[fqn][1] * (
+                        fqn_cofg_dict[fqn][2] + fqn_cofg_dict[fqn][3]
+                    )
+        if flag == 0:
+            _sum += parameter.numel() * 16
+
+    _sum_in_byte = _sum / 8.0
+    _sum_in_GB = _sum_in_byte / (1024**3) / 1.0
+    return _sum_in_GB
+
+# return evaluation results to complete BO trials
+def eval(model, tokenizer, limit, fqn_to_config):
+    return {
+        "cal_PPL": (cal_wikitext_ppl(model, tokenizer, limit), 0.0),
+        "model_size": (cal_model_size(model, fqn_to_config), 0.0),
+    }
+
+def load_model(repo_id, device):
+    tokenizer = AutoTokenizer.from_pretrained(repo_id)
+    model = AutoModelForCausalLM.from_pretrained(repo_id, torch_dtype=torch.bfloat16).to(
+        device=device
+    )
+    return model, tokenizer
+
+def define_parameter_list():
```

Comment:
This looks pretty specific to Llama3. Can we express this in terms of a yaml or json file so users can specify their custom model formats?

Reply:
Made it as a TODO and will support it in the next version